src/basic/io-util.c

```diff
@@ -125,21 +128,36 @@ int loop_write(int fd, const void *buf, size_t nbytes, bool do_poll) {
                 p = buf;
         }
 
+        end = timestamp_is_set(timeout) ? usec_add(now(CLOCK_MONOTONIC), timeout) : USEC_INFINITY;
+
         do {
                 ssize_t k;
+                usec_t t;
+
+                if (end != USEC_INFINITY) {
+                        t = now(CLOCK_MONOTONIC);
+                        if (t >= end)
+                                return -ETIME;
+                } else
+                        t = 0; /* for usec_sub_unsigned below */
 
                 k = write(fd, p, nbytes);
                 if (k < 0) {
                         if (errno == EINTR)
                                 continue;
 
                         if (errno == EAGAIN && do_poll) {
```

Comment:
let's maybe reverse this check do save one level of indentation. ie. something like this:

i.e.:

```c
if (errno != EAGAIN || timeout == 0)
        return -errno;

usec_t t = now(CLOCK_MONOTONIC);
usec_t end = usec_add(start, timeout);
if (t >= end)
        return -ETIME;
r = fd_wait_for_event(fd, POLLOUT, usec_sub_unsigned(end, t));
if (r < 0)
        return r;
if (r == 0)
        return -ETIME;
continue;